frontend/src/components/Pages/FrontPage/MissionOverview/FailedMissionAlertView.tsx

```diff
@@ -100,27 +102,57 @@ export function FailedMissionAlertView({ refreshInterval }: RefreshProps) {
         sessionStorage.setItem(DismissalTimeSessionKeyName, JSON.stringify(Date.now()))
         setRecentFailedMissions([])
     }
 
-    const updateRecentFailedMissions = useCallback(() => {
-        const lastDismissTime: Date = getLastDismissalTime()
-        BackendAPICaller.getMissionRuns({ statuses: [MissionStatus.Failed], pageSize: PageSize }).then((missions) => {
-            const newRecentFailedMissions = missions.content.filter((m) => new Date(m.endTime!) > lastDismissTime)
-            setRecentFailedMissions(newRecentFailedMissions)
-        })
-    }, [PageSize, getLastDismissalTime])
+    // Set the initial failed missions when loading the page or changing installations
+    useEffect(() => {
+        const updateRecentFailedMissions = () => {
+            const lastDismissTime: Date = getLastDismissalTime()
+            BackendAPICaller.getMissionRuns({ statuses: [MissionStatus.Failed], pageSize: PageSize }).then(
+                (missions) => {
+                    const newRecentFailedMissions = missions.content.filter(
+                        (m) =>
+                            new Date(m.endTime!) > lastDismissTime &&
+                            (!installationCode ||
+                                m.installationCode!.toLocaleLowerCase() !== installationCode.toLocaleLowerCase())
+                    )
+                    setRecentFailedMissions(newRecentFailedMissions)
+                }
+            )
+        }
+        if (!recentFailedMissions || recentFailedMissions.length === 0) updateRecentFailedMissions()
+    }, [installationCode])
 
+    // Register a signalR event handler that listens for new failed missions
+    useEffect(() => {
+        if (connectionReady)
+            registerEvent('mission run failed', (username: string, message: string) => {
+                setNewFailedMission(JSON.parse(message))
+                console.log(JSON.parse(message))
+            })
+    }, [registerEvent, connectionReady])
+
+    // Use the failed missions from signalR messages to update the displayed list of failed missions
     useEffect(() => {
-        const id = setInterval(() => {
-            updateRecentFailedMissions()
-        }, refreshInterval)
-        return () => clearInterval(id)
-    }, [refreshInterval, updateRecentFailedMissions])
+        if (newFailedMission) {
+            const lastDismissTime: Date = getLastDismissalTime()
+            if (
+                installationCode &&
+                (!newFailedMission.installationCode ||
+                    newFailedMission.installationCode.toLocaleLowerCase() !== installationCode.toLocaleLowerCase())
+            )
+                return
+            if (new Date(newFailedMission.endTime!) <= lastDismissTime) return
+            let isDuplicate = recentFailedMissions.filter((m) => m.id === newFailedMission.id).length > 0
+            if (isDuplicate) return
+            setRecentFailedMissions([...recentFailedMissions, newFailedMission])
+        }
```

Comment:
```suggestion
        if (newFailedMission) {
            goodFunctionNameHere()
        }
```

I find quite a few of these useEffect functions to be difficult to read without using quite a bit of effort to read them. Could it be an option to wrap each on in a sub function for better readability? It would also make the declarations shorter and thus the entire setup part easier to read.

Reply:
I've removed the need for this useEffect